File: src/main/java/art/chibi/telemetry/TelemetryPlugin.java
package art.chibi.telemetry;

import art.chibi.telemetry.reporter.ErrorReporter;
import art.chibi.telemetry.reporter.ServerStatsReporter;
import org.bukkit.plugin.java.JavaPlugin;

import javax.inject.Inject;
import java.util.logging.Level;

public class TelemetryPlugin extends JavaPlugin {

    @Inject
    ServerStatsReporter serverStatsTelemetry;

    @Inject
    ErrorReporter errorReporter;

    @Override
    public void onEnable() {
        try {
            // 1) Load plugin configuration from config.yml
            Config telemetryConfig = loadConfiguration();

            // 2) Build the Dagger component
            TelemetryComponent component = DaggerTelemetryComponent.builder()
                    .pluginInstance(this)
                    .logger(getLogger())
                    .telemetryConfig(telemetryConfig)
                    .build();

            // 3) Inject this plugin, populating serverStatsTelemetry & errorReporter
            component.inject(this);

            // 4) Install the Log4j2 appender for capturing exceptions
            errorReporter.installAppender();

            // 5) Initialize & schedule periodic stats
            serverStatsTelemetry.init(); // e.g. sets CPU markers
            serverStatsTelemetry.schedule(telemetryConfig.telemetryInterval());

            getLogger().info("TelemetryPlugin activated. Telemetry interval: "
                    + telemetryConfig.telemetryInterval() + " Ticks.");
        } catch (Exception e) {
            getLogger().log(Level.SEVERE, "Initialization error", e);
        }
    }

    private Config loadConfiguration() {
        saveDefaultConfig();

        String hasuraAdminSecret = getConfig().getString("hasura-admin-secret", "");
        String hasuraEndpoint = getConfig().getString("hasura-endpoint", "https://graphql.chibi.art");
        int telemetryInterval = getConfig().getInt("telemetry-interval", 200);
        String metricsTable = getConfig().getString("metrics-table", "minecraft_metrics");
        String exceptionLogTable = getConfig().getString("exception-log-table", "exception_logs");

        return new Config(
                hasuraAdminSecret,
                hasuraEndpoint,
                telemetryInterval,
                metricsTable,
                exceptionLogTable
        );
    }
}

File: src/main/java/art/chibi/telemetry/TelemetryPlugin.kt
package art.chibi.telemetry

import art.chibi.telemetry.config.TelemetryConfig
import art.chibi.telemetry.di.DaggerTelemetryComponent
import art.chibi.telemetry.di.TelemetryComponent
import org.bukkit.plugin.java.JavaPlugin
import java.util.logging.Level
import javax.inject.Inject

class TelemetryPlugin : JavaPlugin() {

    // Injected services
    @Inject
    lateinit var serverStatsTelemetry: ServerStatsTelemetry

    @Inject
    lateinit var errorReporter: ErrorReporter

    override fun onEnable() {
        try {
            // 1) Load plugin configuration from config.yml
            val telemetryConfig = loadConfiguration()

            // 2) Build the Dagger component
            val component: TelemetryComponent = DaggerTelemetryComponent.builder()
                .pluginInstance(this)
                .logger(logger)
                .telemetryConfig(telemetryConfig)
                .build()

            // 3) Inject this plugin, populating serverStatsTelemetry & errorReporter
            component.inject(this)

            // 4) Install the Log4j2 appender for capturing exceptions
            errorReporter.installAppender()

            // 5) Initialize & schedule periodic stats
            serverStatsTelemetry.init()   // e.g. sets CPU markers
            serverStatsTelemetry.schedule(telemetryConfig.telemetryInterval)

            logger.info("TelemetryPlugin activated. Telemetry interval: ${telemetryConfig.telemetryInterval} Ticks.")
        } catch (e: Exception) {
            logger.log(Level.SEVERE, "Initialization error", e)
        }
    }

    private fun loadConfiguration(): TelemetryConfig {
        saveDefaultConfig()

        val hasuraAdminSecret = config.getString("hasura-admin-secret", "") ?: ""
        val hasuraEndpoint = config.getString("hasura-endpoint", "https://graphql.chibi.art")
            ?: "https://graphql.chibi.art"
        val telemetryInterval = config.getInt("telemetry-interval", 200)
        val metricsTable = config.getString("metrics-table", "minecraft_metrics") ?: "minecraft_metrics"
        val exceptionLogTable = config.getString("exception-log-table", "exception_logs") ?: "exception_logs"

        return TelemetryConfig(
            hasuraAdminSecret = hasuraAdminSecret,
            hasuraEndpoint = hasuraEndpoint,
            telemetryInterval = telemetryInterval,
            metricsTable = metricsTable,
            exceptionLogTable = exceptionLogTable
        )
    }
}
